### ISSUE TITLE
Feature Request: Partial Autocorrelation Function (PACF)

### ISSUE THREAD

**Issue by nikhaas** [blue-yonder/tsfresh#223]
For a timeseries with an unknown data generating model, the *autocorrelation function (ACF)* and [*partial autocorrelation function (PACF)*](https://en.wikipedia.org/wiki/Partial_autocorrelation_function) help in identifying the order of an Autoregressive `ARMA(p, q)` model. The [Box-Jenkins method  uses ACF and PACF](https://en.wikipedia.org/wiki/Box%E2%80%93Jenkins_method#Autocorrelation_and_partial_autocorrelation_plots) for this purpose. Depending on the shape of the ACF and PACF, one can derive a model for fitting the data.

The difference between ACF and PACF:

- The ACF calculates the linear relationship between the values at timestep `t` and `t-k` (if one looks at the autocorrelation of lag `k`). So this is an univariate regression problem, which can be solved by Ordinary Least Squares (OLS).
- The PACF calculates the linear relationship between the values at timestep `t` **up to** `t-k` (if one looks at the partial autocorrelation of lag `k`). So this is a **multivariate** regression problem, which can be solved by Ordinary Least Squares (OLS), too. See this [post at SO](https://stats.stackexchange.com/questions/18599/pacf-manual-calculation) for clarification.

#### Current Status

Afaics,`tsfresh` already provides the functions

- [`mean_autocorrelation`](https://github.com/blue-yonder/tsfresh/blob/master/tsfresh/feature_extraction/feature_calculators.py#L264)
- [`autocorrelation`](https://github.com/blue-yonder/tsfresh/blob/master/tsfresh/feature_extraction/feature_calculators.py#L1158)

If I derive it correctly from the statement `"autocorrelation": [{"lag": lag} for lag in range(10)],` from the [`feature_extraction.settings`](https://github.com/blue-yonder/tsfresh/blob/master/tsfresh/feature_extraction/settings.py#L108), the latter is done for the lags 0 (which is trivially 1) to 9.

`tsfresh` also fits an AR process by the method [`ar_coefficient`](https://github.com/blue-yonder/tsfresh/blob/master/tsfresh/feature_extraction/feature_calculators.py#L945). I need to confess that I don't get which params are calculated. It says in the [feature_extraction settings](https://github.com/blue-yonder/tsfresh/blob/master/tsfresh/feature_extraction/settings.py#L117)
```python
"ar_coefficient": [{"coeff": coeff, "k": k} for coeff in range(5) for k in [10]],
```
So `coeff` stands for the number of coefficients, i.e. fitting an `AR(coeff)` model, but what does `k` stand for? Could one decide to fit several AR-processes, whose parameter settings could be identified by the ID `k`? Am I getting this right?
If so, then `tsfresh` fits an arbitrary defined AR-model.

#### Include PACF values
Including the PACF values, e.g. by using the [`statsmodels.tsa.stattools.pacf`](http://www.statsmodels.org/dev/generated/statsmodels.tsa.stattools.pacf.html#statsmodels.tsa.stattools.pacf), would reveal some imformation that can be exploited by a subsequent classifier. I recommend using the `method='ols'` parameter. Here is a small example:

- The AR(3) model:
```python
ar_process = [1] + [3.5] + [8.25] + 47 * [0]  # 50 data points, first 3 calculated by hand
for i in range(3, len(ar_process)):
    ar_process[i] = \
    2.5 * ar_process[i - 1] + \
    -1.5 * ar_process[i - 2] + \
    0.5 * ar_process[i - 3] + \
    1
print(ar_process[:10])
# > [1, 3.5, 8.25, 16.875, 32.5625, 61.21875, 113.640625, 209.5546875, 385.03515625, 706.076171875]
```
- PACF and ACF with 49 lags
```python
lags = len(ar_process)-1
f, axarr = plt.subplots(1, 2, figsize=(20,5))
plot_pacf(ar_process, method='ols', lags=lags, ax=axarr[0].axes)
plot_acf(ar_process, lags=lags, ax=axarr[1].axes)
plt.suptitle('lags: {}'.format(lags))
plt.show()
```
![pacf_acf_lags_49](https://user-images.githubusercontent.com/18031651/27803095-16e734ac-6028-11e7-8e2b-c052a522c340.png)
- PACF and ACF with 10 lags

```python
lags = 10
f, axarr = plt.subplots(1, 2, figsize=(20,5))
plot_pacf(ar_process, method='ols', lags=lags, ax=axarr[0].axes)
plot_acf(ar_process, lags=lags, ax=axarr[1].axes)
plt.suptitle('lags: {}'.format(lags))
plt.show()
```
![pacf_acf_lags_10](https://user-images.githubusercontent.com/18031651/27803112-2ce0f932-6028-11e7-8cd5-1ff0c7bb7476.png)

- PACF: `method='ols'` vs `method='ywu'`

```python
lags = 10
f, axarr = plt.subplots(1, 2, figsize=(20,5))
plot_pacf(ar_process, method='ols', lags=lags, ax=axarr[0].axes)
plot_pacf(ar_process, method='ywunbiased', lags=lags, ax=axarr[1].axes)
plt.suptitle('ols vs ywunbiased')
plt.show()
```
![pacf_ols_vs_yw](https://user-images.githubusercontent.com/18031651/27803114-33a32ec0-6028-11e7-9be8-b711a337141f.png)

The method on the right side uses the [Yule-Walker equations](https://en.wikipedia.org/wiki/Autoregressive_model#Yule.E2.80.93Walker_equations), which produces fundamentally different results than the `ols` approach and does not estimate the coefficients of the "normal" AR(p) formulation, at least in this example.

**Comment by MaxBenChrist** [comment]
Again, great post, plots, everything great quality.

You could consider to submit that plots in a small notebook as a PR!

> So coeff stands for the number of coefficients, i.e. fitting an AR(coeff) model, but what does k stand for? Could one decide to fit several AR-processes, whose parameter settings could be identified by the ID k? Am I getting this right?

`k` is the max lag parameter of the AR model, `coeff` the index of the parameter that is returned as feature. so it fits a AR(k) model and returns the `coeff <= k` coefficient. if that was not clear or confusing to you. could you please submit a pr where you correct that docstring? 

next steps:

you propose to add the PACF as features. I strongly agree. Do you want to implement them?

**Comment by MaxBenChrist** [comment]
I think we are already calculating the pcaf with the `autocorrelation` feature calculator.

**Comment by nikhaas** [comment]
The `autocorrelation` feature calculator uses `pd.Series.autocorr()`.

I played around in [this notebook. starting from cell 14](https://github.com/nikhaas/tsfresh/blob/acf_pacf_autocorr_notebook/notebooks/autocorr_acf_pacf_comparison.ipynb) to check the interdependence of `autocorr,` and `statsmodels acf, pacf`. For the simple AR(1) process

```python
np.random.seed(0)
T = 15
c = 2
epsilon = np.random.randn(T)
ar_process = np.repeat(1.0, T)
phi = 0.85
for t in range(T-1):
    ar_process[t+1] = phi*ar_process[t] + c  + epsilon[t]

ar_process = pd.Series(ar_process)
``` 
which looks like
![simple_ar_process](https://user-images.githubusercontent.com/18031651/28684891-53b0b9f8-7306-11e7-8749-e5ed2679aef5.png)

`autocorr, acf, pacf` look like 
![autocorr_vs_acf_vs_pacf](https://user-images.githubusercontent.com/18031651/28684915-672a7f5a-7306-11e7-87ad-16e22e3deddc.png)

I draw the following conclusion (see also in the [Notebook](https://github.com/nikhaas/tsfresh/blob/acf_pacf_autocorr_notebook/notebooks/autocorr_acf_pacf_comparison.ipynb)):

## Conclusion
- There are no tests for autocorrelation atm, maybe one should add some
- Tbh, I expected the values for autocorr and acf to be the same. What am I missing?
- I tried to match the autocorr results with the acf results, but I failed.
- ACF and PACF generate values outside the value range [-1, 1]. One could catch this as exceptions
- autocorr is using np.corrcoef internally.
- The values of autocorr, acf and pacf are different except for lag 1 but are correlated in general
- Shall PACF be included in tsfresh, too?

## PACF Issue

I think I was oversimplifying things in my first post. I haven't understood how ACF and PACF work yet. 
However, I [opened an Issue in statsmodels](https://github.com/statsmodels/statsmodels/issues/3839), because I think their PACF with OLS is wrong.

## TL;DR
> I think we are already calculating the pcaf with the autocorrelation feature calculator.

- I don't think so. A plot shows different but correlated results.
- A test for `autocorrelation` is missing, maybe write a dummy
- Decide to use PACF as a Black Box and trust the statsmodels API with default settings or disregard it

**Comment by MaxBenChrist** [comment]
> There are no tests for autocorrelation atm, maybe one should add some

Agree. 

> I don't think so. A plot shows different but correlated results.

Well. ok, then lets include the PCAF as well.

Can you provide prs (a) for the PCAF and (b) with unit tests for the autocorrelation function, Niklas?

**Comment by MaxBenChrist** [comment]
> ACF and PACF generate values outside the value range [-1, 1]. One could catch this as exceptions

I would not do that.

**Comment by MaxBenChrist** [comment]
I moved the investigation regarding the acf into a new issue #273

**Comment by nikhaas** [comment]
> Can you provide prs (a) for the PCAF and (b) with unit tests for the autocorrelation function, Niklas?

Yes, I can do this. This may take a while, I hope to finish it within a few days.
Moreover, I can suggest to refine the notebook to present the features `autocorr, acf, pacf` if desired. 
I suggest 3 PRs: 
(a) for the autocorr tests 
(b) for the PACF function with tests and 
(c) for the explanatory notebook.

I made a quick speed comparison ( [happy to save time by modifying your code from the DFT issue :-)](https://github.com/blue-yonder/tsfresh/issues/228)):

```python
import numpy as np
import pandas as pd
import matplotlib.pylab as plt
import time
from statsmodels.tsa.stattools import pacf, acf

lengths = [10, 50, 100, 200, 300, 400, 500, 600, 700, 1000, 2000, 3000, 4000, 5000, 10000]
t_yw = [0]*len(lengths)
t_ld = [0]*len(lengths)
t_acf = [0]*len(lengths)

n_iter = 100

for _ in range(n_iter):
    i=0
    for l in lengths:
        x = np.random.normal(size=l)
        start_time = time.time()
        pacf(x, nlags=min(len(x), 100)-1, method='ywunbiased')
        t_yw[i] += (time.time() - start_time)
        i+=1

for _ in range(n_iter):
    i=0                             
    for l in lengths:               
       x = np.random.normal(size=l)
       start_time = time.time()                           
       pacf(x, nlags=min(len(x), 100)-1, method='ld')
       t_ld[i] += (time.time() - start_time)
       i+=1 

for _ in range(n_iter):
    i=0                             
    for l in lengths:               
       x = np.random.normal(size=l)
       start_time = time.time()                           
       acf(x, nlags=min(len(x), 100)-1, unbiased=True)
       t_acf[i] += (time.time() - start_time)
       i+=1 

plt.plot(lengths, t_yw, "o-", label="yw")
plt.plot(lengths, t_ld, "o-", label="ld")
plt.plot(lengths, t_acf, "o-", label="acf")
plt.legend(["yw", "ld", "acf"])
plt.xlabel("Individual time series length")
plt.ylabel("Total runtime in seconds for {} time series".format(n_iter))
plt.show()

```
Result:
![yw_vs_ld_vs_acf](https://user-images.githubusercontent.com/18031651/28688965-25cb812c-7314-11e7-8dcd-3abfec66180c.png)

The feature is not that cheap, but the overhead seems to be minimal compared to `acf` which is already used in a tsfresh feature calculator. Of course, I am going to use the `method="ld"` parameter ( which enables [Levinson_recursion](https://en.wikipedia.org/wiki/Levinson_recursion) instead of [Yule-Walker Equations](https://en.wikipedia.org/wiki/Autoregressive_model#Yule.E2.80.93Walker_equations)).

**Comment by MaxBenChrist** [comment]
> The feature is not that cheap, but the overhead seems to be minimal compared to acf which is already used in a tsfresh feature calculator. Of course, I am going to use the method="ld" parameter ( which enables Levinson_recursion instead of Yule-Walker Equations).

The yw equations look to slow, 10 seconds for one time series of length 1000.... However, we use random time series, maybe it is a little bit faster if you have a time series with "structure".

**Comment by nikhaas** [comment]
`acf` means the `statsmodels.tsa.stattools.acf` function. tsfresh already uses this in the feature calculator `agg_autocorrelation`. With this plot I just wanted to point out that the performance of `pacf` with `ld` is close to a function that is already used in tsfresh. `acf` is not an option we can use in the `pacf` function.

It is 10 seconds for 100 timeseries (label on the y axis), therefore it is 0.1 seconds for one time series.
